src/EventGridExtension/EventGridExtensionConfig.cs

```diff
@@ -20,6 +21,19 @@ public class EventGridExtensionConfig : IExtensionConfigProvider,
                        IAsyncConverter<HttpRequestMessage, HttpResponseMessage>
     {
         private ILogger _logger;
+        private readonly Func<EventGridAttribute, IAsyncCollector<EventGridEvent>> _converter;
+
+        // for end to end testing
+        public EventGridExtensionConfig(Func<EventGridAttribute, IAsyncCollector<EventGridEvent>> converter)
+        {
+            _converter = converter;
```

Comment:
Since this is a ctor for testing only, it shouldn't be public on this public type. Make it internal.

Actually, a cleaner way to do what you're trying to do here would be to define the converter as a protected virtual method, and have your test derive a config and override it. You can then make just the client creation portion overridable, to return your mock client.

Reply:
What do you think about the virtual method approach?

Reply:
I now remember why I didn't go with that approach, we cannot mock `EventGridExtensionConfig` because the sdk will try to get attributes from the extension assembly and it won't work with moq:
![image](https://user-images.githubusercontent.com/6531400/42713479-7a524cb8-86a4-11e8-8e2b-aaa2ff31d3ac.png)

Reply:
You don't have to moq the config - you just derive a test config from the base and override the one method to create the test factory.

However, if you don't want to go that route, I'd say just change your internal test ctor to take an EventGridClient instance, rather than the converter func, so your mock remains more targeted. Would that work?